onmt/utils/optimizers.py

```diff
@@ -2,87 +2,108 @@
-            decay_method=opt.decay_method,
-            warmup_steps=opt.warmup_steps,
-            model_size=opt.rnn_size)
-
-    # Stage 1:
-    # Essentially optim.set_parameters (re-)creates and optimizer using
-    # model.paramters() as parameters that will be stored in the
-    # optim.optimizer.param_groups field of the torch optimizer class.
-    # Importantly, this method does not yet load the optimizer state, as
-    # essentially it builds a new optimizer with empty optimizer state and
-    # parameters from the model.
-    optim.set_parameters(model)
-
-    if opt.train_from and (opt.reset_optim in ['none', 'keep_states']):
-        # Stage 2: In this stage, which is only performed when loading an
-        # optimizer from a checkpoint, we load the saved_optimizer_state_dict
-        # into the re-created optimizer, to set the optim.optimizer.state
-        # field, which was previously empty. For this, we use the optimizer
-        # state saved in the "saved_optimizer_state_dict" variable for
-        # this purpose.
-        # See also: https://github.com/pytorch/pytorch/issues/2830
-        optim.optimizer.load_state_dict(saved_optimizer_state_dict)
-        # Convert back the state values to cuda type if applicable
-        if use_gpu(opt):
-            for state in optim.optimizer.state.values():
-                for k, v in state.items():
-                    if torch.is_tensor(v):
-                        state[k] = v.cuda()
-
-        # We want to make sure that indeed we have a non-empty optimizer state
-        # when we loaded an existing model. This should be at least the case
-        # for Adam, which saves "exp_avg" and "exp_avg_sq" state
-        # (Exponential moving average of gradient and squared gradient values)
-        if (optim.method == 'adam') and (len(optim.optimizer.state) < 1):
-            raise RuntimeError(
-                "Error: loaded Adam optimizer from existing model" +
-                " but optimizer state is empty")
-
-    return optim
+        raise ValueError('Invalid optimizer type: ' + opt.optim)
+    return optimizer
+
+
+def make_learning_rate_decay_fn(opt):
+    """Returns the learning decay function from options."""
+    if opt.decay_method == 'noam':
+        return noam_decay_fn(opt.warmup_steps, opt.rnn_size)
+    elif opt.start_decay_steps is not None:
+        return exponential_decay_fn(
+            opt.learning_rate_decay, opt.decay_steps, opt.start_decay_steps)
+
+
+def noam_decay_fn(warmup_steps, model_size):
+    """Learning rate schedule described in
+    https://arxiv.org/pdf/1706.03762.pdf.
+    """
+    return lambda step: (
+        model_size ** (-0.5) *
+        min(step ** (-0.5),
+            step * warmup_steps**(-1.5)))
+
+
+def exponential_decay_fn(rate, decay_steps, start_step=0):
+    """A standard exponential decay, scaling the learning rate by :obj:`rate`
+    every :obj:`decay_steps` steps.
+    """
+    return lambda step: (
```

Comment:
Is there a case where an `Optimizer` object might need to be serialized? If so, I think returning a lambda here might be a problem because this lambda expression will end up as an attribute of the `Optimizer`. The only advantage lambdas have over other ways of defining functions is that they can give you short one-liners; other than that, they have only disadvantages. The code here looks like it could be rewritten with `functools.partial`, which is serializable.

Reply:
Or just with normal functions.